File: ArrayAndString/reverse-words.js
/*Given an input string s, reverse the order of the words.
A word is defined as a sequence of non-space characters. The words in s will be separated by at least one space.*/

const reverseWords = (s) => {
    let wordArr = s.split(' ');
    for (let i = 0; i < wordArr.length; i++) {
        console.log(`"${wordArr[i]}"`);
        if (wordArr[i] === '') {
            wordArr = wordArr.slice(0, i).concat(wordArr.slice(i + 1));
            i--;
        }
    }

    return wordArr.reverse().join(' ');
    
}

//O(n^2)
/*	The code has a loop that iterates through each word in the input string, splitting it into an array. 
Inside the loop, there is a conditional check and potentially a slice operation which can take O(n) time. 
As a result, the overall time complexity is O(n^2) 
where n is the length of the input string.*/

const reverseWordsTrimRegex = (s) => s.trim().split(/\s+/).reverse().join(' ');

//O(n)
/*The code first trims the input string, which takes O(n) time where n is the length of the input string. 
Then it splits the string into an array of words using the split method, which also takes O(n) time. 
The reverse method reverses the array in place, which is O(n) as well. 
Finally, the join method concatenates the reversed array into a string, which also takes O(n) time. 
Therefore, the overall time complexity is O(n).*/

//doesn't use built-in methods such as split and trim
const reverseWordsDeque = (s) => {
    let left = 0,
        right = s.length - 1;
    
    //for removing/not dealing with leading and trailing spaces (instead of trim)
    while (left <= right && s[left] === ' ') left++;
    while (left <= right && s[right] === ' ') right--;

    let dequeu = [];
    let word = '';

    while (left <= right) {
        if (word.length !== 0 && s[left] === ' ') {
            dequeu.push(word);
            word = '';
        } else if (s[left] !== ' ') {
            word += s[left];
        }
        left++;
    }

    dequeu.push(word);

    return dequeu.reverse().join(' ');
}

//O(n)
/*The code iterates through the input string 's' once to extract words and store them in a deque. 
The final step of reversing and joining the words in the deque has a time complexity of O(n) as well. 
Therefore, the overall time complexity is O(n).*/

const s = " a good   example ";
console.log(reverseWordsDeque(s));

File: ArrayAndString/reverse_words.py
class ReverseWords:
    def reverseWordsManualReversal(self, s: str) -> str:
        word_list = s.split()
        left, right = 0, len(word_list) - 1
        while left < right:
            word_list[left], word_list[right] = word_list[right], word_list[left]
            left += 1
            right -= 1

        return " ".join(word_list)

    def reverseWordsBuiltIn(self, s: str) -> str:
        return " ".join(reversed(s.split()))

    # Time complexity: O(n) where n is the length of the input string s.
    # We split the string into words, which takes O(n) time,
    # and then we reverse the list of words in-place,
    # which also takes O(n) time in the worst case (when all characters are part of a single word).
    # Finally, we join the words back into a string, which takes O(n) time.
    # Space complexity: O(n) for the list of words created by splitting the input string.

    def reverseWordsFollowUp(self, s: str) -> str:
        def reverse(left, right):
            while left < right:
                s[left], s[right] = s[right], s[left]
                left += 1
                right -= 1

        # Convert string to list for in-place modifications
        s = list(s)
        n = len(s)

        # 1. Reverse entire string
        reverse(0, n - 1)

        # 2. Reverse each word
        left = 0
        while left < n:
            if s[left] == " ":
                left += 1
                continue

            right = left
            while right < n and s[right] != " ":
                right += 1

            reverse(left, right - 1)
            left = right

        # 3. Clean spaces in-place
        write = 0
        read = 0

        while read < n:
            # skip spaces
            while read < n and s[read] == " ":
                read += 1

            # copy word
            while read < n and s[read] != " ":
                s[write] = s[read]
                write += 1
                read += 1

            # skip spaces between words
            while read < n and s[read] == " ":
                read += 1

            # add single space if more words remain
            if read < n:
                s[write] = " "
                write += 1

        # trim extra characters
        return "".join(s[:write])
